src/main/java/com/shop/service/ProductServiceImpl.java

```diff
@@ -1,17 +1,31 @@
 package com.shop.service;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.shop.dto.ProductDto;
+import com.shop.dto.SpecificationDto;
+import com.shop.model.Image;
 import com.shop.model.Product;
+import com.shop.model.ProductSpecification;
+import com.shop.model.Specification;
 import com.shop.repository.ProductRepository;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.io.IOException;
+import java.util.*;
+
 @Service
 public class ProductServiceImpl implements ProductService {
 
     @Autowired
     ProductRepository productRepository;
 
+    @Autowired
+    DtoService dtoService;
+
+    @Autowired
```

Comment:
isn't better to inject by constructor here>

Reply:
sometimes injecting by constructor may make a loop, will check it